farm-staking/farm-staking/src/lib.rs

```diff
@@ -306,47 +306,90 @@ pub trait Farm:
     fn claim_rewards(&self) -> ClaimRewardsResultType<Self::Api> {
         let payments = self.call_value().all_esdt_transfers();
 
-        self.claim_rewards_common(payments, None)
-    }
+        require!(self.is_active(), "Not active");
+        require!(!self.farm_token_id().is_empty(), "No farm token");
 
-    #[payable("*")]
-    #[endpoint(claimRewardsWithNewValue)]
-    fn claim_rewards_with_new_value(
-        &self,
-        new_values: ManagedVec<BigUint>,
-    ) -> ClaimRewardsResultType<Self::Api> {
-        let caller = self.blockchain().get_caller();
-        self.require_whitelisted(&caller);
+        let first_payment = payments
+            .try_get(0)
+            .unwrap_or_else(|| sc_panic!("empty payments"));
+        let additional_payments = payments.slice(1, payments.len()).unwrap_or_default();
 
-        let payments = self.call_value().all_esdt_transfers();
-        require!(
-            payments.len() == new_values.len(),
-            "Arguments length mismatch"
+        let payment_token_id = &first_payment.token_identifier;
+        let old_farming_amount = &first_payment.amount;
+        let token_nonce = first_payment.token_nonce;
+
+        require!(old_farming_amount > &0u32, "Zero amount");
+        let farm_token_id = self.farm_token_id().get();
+        require!(payment_token_id == &farm_token_id, "Unknown farm token");
+        let farm_attributes = self
+            .get_attributes::<StakingFarmTokenAttributes<Self::Api>>(payment_token_id, token_nonce);
+
+        let reward_token_id = self.reward_token_id().get();
+        self.generate_aggregated_rewards();
+
+        let current_reward_per_share = self.reward_per_share().get();
+        let reward = self.calculate_reward(
+            old_farming_amount,
+            &current_reward_per_share,
+            &farm_attributes.reward_per_share,
+        );
+        let new_compound_reward_amount = self.rule_of_three(
```

Comment:
compounded reward should not change by the claim rewards. I don't see a reason to compute it again.